ibridges/executor.py

```diff
@@ -0,0 +1,440 @@
+"""Operations to be performed for upload/download/sync."""
+from __future__ import annotations
+
+import json
+import warnings
+from collections import defaultdict
+from pathlib import Path
+from typing import Optional, Union
+
+import irods.collection
+import irods.data_object
+import irods.exception
+import irods.keywords as kw
+from tqdm import tqdm
+
+from ibridges.export_metadata import add_to_metadict, empty_metadict, set_metadata_from_dict
+from ibridges.path import IrodsPath
+from ibridges.session import Session
+
+NUM_THREADS = 4
+
+
+class Operations():  # pylint: disable=too-many-instance-attributes
+    """Storage for all data and metadata operations.
+
+    This class is mostly for internal use.
+    """
+
+    def __init__(self, resc_name: Optional[str] = None, options: Optional[dict] = None):
+        """Initialize and empty Operations object.
+
+        Parameters
+        ----------
+        resc_name:
+            Name of the resource to perform the operations on.
+        options:
+            Options to transfer data with.
+
+        """
+        self.create_dir: set[str] = set()
+        self.create_collection: set[str] = set()
+        self.upload: list[tuple[Path, IrodsPath]] = []
+        self.download: list[tuple[IrodsPath, Path]] = []
+        self.meta_download: dict = defaultdict(lambda: {"items": []})
+        self.meta_upload: list[tuple[IrodsPath, Union[str, Path]]] = []
+        self.resc_name: str = "" if resc_name is None else resc_name
+        self.options: Optional[dict] = {} if resc_name is None else options
+
+    def add_meta_download(self, root_ipath: IrodsPath, ipath: IrodsPath, meta_fp: Union[str, Path]):
```

Comment:
I would turn around the logic. There should be a metadata download function which can be triggered by the general download function. This will allow to only download metadata.